File: ocaml/lib/s07.ml
type entry = Dir of (string, entry) Hashtbl.t | File of int

let rec pp_entry fmt t =
  match t with
  | Dir entries ->
      entries |> Hashtbl.to_seq |> List.of_seq |> List.sort compare
      |> pp_tree fmt
  | File s -> Format.fprintf fmt "(size=%d)" s

and pp_tree fmt l =
  List.iter
    (fun (name, e) -> Format.fprintf fmt "@[- %s @[%a@]@\n" name pp_entry e)
    l

let iter_on_dir_size f t =
  let rec loop t =
    match t with
    | File s -> s
    | Dir entries ->
        let size = Hashtbl.fold (fun _ e acc -> acc + loop e) entries 0 in
        f size; size
  in
  ignore (loop t)

let sum_size limit t =
  let acc = ref 0 in
  let f s = if s <= limit then acc := !acc + s in
  iter_on_dir_size f t; !acc

let find_dir total needed t =
  let acc = ref [] in
  let f size = acc := size :: !acc in
  iter_on_dir_size f t;
  match !acc with
  | root_size :: rest ->
      let available = total - root_size in
      assert (available < needed);
      let l = List.sort compare rest in
      List.find (fun s -> s + available >= needed) l
  | _ -> assert false

let mk_dir () = Hashtbl.create 16

let solve compute () =
  let root_dir = mk_dir () in
  let root_path = [ "/", root_dir ] in
  let _path =
    Utils.fold_fields ' '
      (fun path cmd ->
        match path, cmd with
        | _, [ "$"; "cd"; "/" ] -> root_path
        | _ :: parent_path, [ "$"; "cd"; ".." ] -> parent_path
        | (_, cwd) :: _, [ "$"; "cd"; name ] -> begin
            if not (Hashtbl.mem cwd name) then
              Hashtbl.add cwd name (Dir (mk_dir ()));
            match Hashtbl.find cwd name with
            | File _ -> failwith "Cannot cd into a file"
            | Dir d -> (name, d) :: path
          end
        | _, [ "$"; "ls" ] -> path
        | _, [ "dir"; _ ] -> path
        | (_, cwd) :: _, [ ssize; name ] ->
            Hashtbl.replace cwd name (File (int_of_string ssize));
            path
        | _ -> assert false)
      root_path
  in
  let r = mk_dir () in
  Hashtbl.add r "/" (Dir root_dir);
  let t = Dir r in
  Format.printf "%d@\n" (compute t)

module Sol = struct
  let name = "07"
  let solve_part1 = solve (sum_size 100000)
  let solve_part2 = solve (find_dir 70000000 30000000)
end

let () = Solution.register_mod (module Sol)

File: python/s07.py
from sys import stdin

def load_commands():
    root = ("/", {})
    path = [root]
    for line in stdin:
        cmd = line.strip().split(" ")
        if len(cmd) == 0:
            continue
        if cmd[0] == "$" and cmd[1] == "ls":
            pass
        elif cmd[0] == "$" and cmd[1] == "cd":
            dest = cmd[2]
            if dest == "/":
                path = [root]
            elif dest == "..":
                path.pop()
            else:
                # dest is an explicit directory
                parent = path[-1]  # last element
                if dest not in parent[1]:
                    parent[1][dest] = {}
                path += [(dest, parent[1][dest])]
        else:
            cwd = path[-1]
            arg = cmd[1]
            if cmd[0] == "dir":
                if arg not in cwd[1]:
                    cwd[1][arg] = {}
            else:
                cwd[1][arg] = int(cmd[0])

    return {"/":root[1]}

##debugging
def pr_tree(dir, depth=0):
    for k in dir.keys():
        print(depth * " ", "-", k, end='')
        if type(dir[k]) is int:
            print(f" (file, size={dir[k]})")
        else:
            print(" (dir)")
            pr_tree(dir[k], depth=depth+1)


def size_dir(dir, res):
    total = 0
    for k in dir.keys():
        v = dir[k]
        if type(v) is int:
            dsize = v
        else:
            dsize = size_dir(v, res)
            res += [(k, dsize)]
        total += dsize
    return total

def solve_part1():
    d = load_commands()
    res = []
    root_size = size_dir(d, res)
    total = 0
    for (_, s) in res:
        if s <= 100000:
            total+= s
    print(total)

def solve_part2():
    d = load_commands()
    res = []
    root_size = size_dir(d, res)
    available = 70000000 - root_size
    needed = 30000000 - available
    if needed <= 0:
        print ("Impossible", available, root_size)
        return
    min_dir = 30000000
    for (_, s) in res:
        if s >= needed and s < min_dir:
            min_dir = s
    print(min_dir)